### DISCUSSION TITLE
RenderTexture error

### DISCUSSION THREAD

**Discussion by 10100010010001001010** [axmolengine/axmol#1097] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/120378517/223126799-683a8bab-61db-41eb-acb9-299c10cab625.png)

I get this error after adding below code:

`auto sprite1 = Sprite::create("spaceShip.png");
    sprite1->setScale(0.5); 
    sprite1->setPosition(visible.width / 2 - visible.width / 4, visible.height / 2); //child sprite - ship
    this->addChild(sprite1, 1);

    auto scaled = RenderTexture::create(ship->getContentSize().width * 0.5, ship->getContentSize().height * 0.5, PixelFormat::RGB8);

    scaled->begin();
    sprite1->visit();
    scaled->end();

    auto scaledSprite = Sprite::createWithTexture(scaled->getSprite()->getTexture());
    scaledSprite->setPosition(visible.width / 2, visible.height / 2);
    this->addChild(scaledSprite);`

How to use this class correctly?

**Comment by aismann** [comment]
> How to use this class correctly?

Please have a look on `cpp-tests`.

**Comment by rh101** [comment, marked as answer]
If you check the location of that crash via a debugger, you would see that the crash is from this assert:
![image](https://user-images.githubusercontent.com/8603230/223347494-10d4f303-1ba3-43c3-819f-29a7c3672e21.png)

It is pointing out exactly what the problem is.

Also, given that you haven't provided any context to your code snippet, we cannot tell where and when you're calling that code.  My guess is that you're using it in the `init()` method, which is **not** the right place for it.  The scene would not have finished initializing at that point.

The way to fix it is to ensure that you do not call that code until after the scene has finished initializing, and there are many ways to handle this.

For example, you can still use it in `init()`, but you need to schedule it for the **next** cycle using `scheduleOnce`, like this:
```
scheduleOnce([this, visibleSize](float) {
    auto* scaledRenderTexture = RenderTexture::create(sprite->getContentSize().width * 0.5, sprite->getContentSize().height * 0.5, PixelFormat::RGB8);

    scaledRenderTexture->beginWithClear(0, 1, 0, 0);
    sprite->setScale(0.5f);
    sprite->setAnchorPoint(Vec2::ANCHOR_BOTTOM_LEFT);
    sprite->setPosition(Vec2(0, 0));
    sprite->visit();
    scaledRenderTexture->end();
    //Director::getInstance()->getRenderer()->render();

    scaledRenderTexture->setPosition(visibleSize.width / 2 + 300, visibleSize.height / 2);
    auto* scaledSprite = Sprite::createWithTexture(scaledRenderTexture->getSprite()->getTexture());
    scaledSprite->setPosition(visibleSize.width / 2, visibleSize.height / 2);
    this->addChild(scaledSprite);
}, 0.0f, "RT");
```

You can also do it in an override of `onEnter()`:

```
void HelloWorld::onEnter()
{
    Scene::onEnter();

    auto* scaledRenderTexture = RenderTexture::create(sprite->getContentSize().width * 0.5, sprite->getContentSize().height * 0.5, PixelFormat::RGB8);

    scaledRenderTexture->beginWithClear(0, 1, 0, 0);
    sprite->setScale(0.5f);
    sprite->setAnchorPoint(Vec2::ANCHOR_BOTTOM_LEFT);
    sprite->setPosition(Vec2(0, 0));
    sprite->visit();
    scaledRenderTexture->end();
    //Director::getInstance()->getRenderer()->render();

    auto visibleSize = _director->getVisibleSize();
    scaledRenderTexture->setPosition(visibleSize.width / 2 + 300, visibleSize.height / 2);
    auto* scaledSprite = Sprite::createWithTexture(scaledRenderTexture->getSprite()->getTexture());
    scaledSprite->setPosition(visibleSize.width / 2, visibleSize.height / 2);
    this->addChild(scaledSprite);
}
```

The other issue is that you're creating a render texture that is half the size of the sprite, yet you're trying to render the full sprite to the render texture, and you haven't set the position of that sprite correctly either within the render texture boundary.  As @aismann mentioned, the cpp-tests project is the best place to learn how to use this and any other functionality within this game engine.
